### DISCUSSION TITLE
Chiamata api, response null

### DISCUSSION THREAD

**Discussion by jehol75** [fattureincloud/api#61] [Q&A] [answered]
Chiamando le API dell' SDK restituisce null

A parte che i dati sembrerebbero corretti, il path e' impostato ma anche tutto questo non fosse corretto/impostato perche' non restituiscono alcun messaggio?

Ci son volute oltre 2 settimane dall' assistenza per sentirsi dire  di scrivere qui, speriamo sia la volta buona

Video allegato

https://user-images.githubusercontent.com/2175833/186106799-dd0cec93-c98c-4cc2-9cc1-bbb91a97fdb1.mp4

**Comment by emajo** [comment]
Ciao @jehol75, il codice che si vede nel video sembrerebbe corretto, per provare a replicare il tuo problema ti chiederei di mandare anche la restante parte del codice in quanto dai nostri log non arrivano proprio le chiamate dalla sua integrazione.

Grazie,
Emanuele

**Reply by jehol75** [reply]
Carissimi, ripeto il problema/domanda

PERCHE' RESTITUSICE NULL E NON UN MESSAGGIO DI ERRORE?

Se volete il codice devo prima chiedere al cliente se e' disposto a sobbarcarsi le spese per mettere a posto i vs problemi.

**Reply by emajo** [reply]
Ciao @jehol75, mi dispiace che stia riscontrando questi problemi, tuttavia la nostra SDK C# viene continuamente utilizzata da migliaia di clienti senza alcun problema, perciò non mi lascia che pensare che effettivamente il problema sia dovuto a un utilizzo non corretto da parte sua.
Come le dicevo prima, la sua chiamata non sta arrivando ai nostri sistemi, di conseguenza ovviamente non può venir restituito nessun messaggio di errore, il null a cui lei si riferisce è il valore dell'oggetto contenente la risposta della nostra api, che ovviamente non viene popolato dato che l'api non viene chiamata.
Noi possiamo aiutarla a risolvere il suo problema se lei si dimostra disponibile, ma, se non dovesse trovarsi a suo agio nell'utilizzo della nostra libreria può sempre utilizzare un client http per fare le chiamate in modo da non dover interfacciarsi con il codice scritto da noi.

Le chiedo inoltre se può evitare di scrivere in CAPS LOCK nelle nostra community in quanto non propriamente consono in una discussione pubblica con persone disposte a impiegare del tempo per aiutarla.

Grazie,
Emanuele

**Reply by jehol75** [reply]
Per l' ennesima volta, non metto in dubbio che possa essere un mio errore se anche non ho capito quale.

"Come le dicevo prima, la sua chiamata non sta arrivando ai nostri sistemi, di conseguenza ovviamente non può venir restituito nessun messaggio di errore" questa frase non ha senso perche' nel momento in cui entro in un vs oggetto se c'e' qualcosa che non va dovrebbe almeno resitutire un exception generica e invece restituisce null

"Le chiedo inoltre se può evitare di scrivere in CAPS LOCK nelle nostra community in quanto non propriamente consono in una discussione pubblica con persone disposte a impiegare del tempo per aiutarla." evitiamo polemiche inutili perche' il tempo lo sto impiegando io per gestire un vostro evidente problema, siete voi che avete chiesto di scrivere qui, per me andava bene andare avanti nella chat interna 

dopo 2 settimana la gente si stanca anche

Inutile partire dal fatto che "e' utlizzata da milioni di utenti" quindi teoricamente perfetta  e' la vostra api che restituisce null come da video, non un fotomontaggio...

comuqnue a breve vi metto codice e json qui (se anche l' avevo gia' inviato nella chat interna) poi vediamo di arrivarci in fondo e il mio caps lock lo usero' da un' altra parte

Saluti

**Reply by jehol75** [reply]
```
   try
            {
                Configuration config = new Configuration();
                config.BasePath = "https://api-v2.fattureincloud.it";
                config.AccessToken = access_token;

                var apiInstance = new IssuedDocumentsApi(config);


                string note = "";

                if (fattura_cliente.FK_azienda == 2)
                {
                    note += "Grazie per la preferenza accordataci.È un piacere lavorare con voi al vostro progetto.Cordiali saluti, SINFONIA cooperativa sociale";
                }
                else
                {
                    note += "Grazie per la preferenza accordataci. È un piacere lavorare con voi al vostro progetto.";
                    note += "Cordiali saluti, AeC Arc en ciel soc coop ";
                }

                string id_metodo_pag = new FattureBL().GetMeotodoPagamentoByID(fattura_cliente.FK_metodo_pagamento,
                    fattura_cliente.FK_MAIN_TIPO_PAGAMENTO, fattura_cliente.FK_azienda);

                DateTime dueDate = fattura_cliente.data_creazione.AddDays(fattura_cliente.scadenza_gg);


                Entity entity = new Entity(
                   id: null,
                   name: fattura_cliente.ClienteDescrizione,
                   vatNumber: fattura_cliente.cliente.piva,
                   taxCode: fattura_cliente.cliente.cf,
                   addressStreet: fattura_cliente.cliente.indirizzo,
                   addressPostalCode: fattura_cliente.cliente.cap,
                   addressCity: fattura_cliente.provincia,
                   addressProvince: fattura_cliente.provincia,
                   country: UppercaseFirst(fattura_cliente.cliente.nazione)
               );

                IssuedDocument invoice = new IssuedDocument(
                    type: IssuedDocumentType.Invoice,
                    entity: entity,
                    date: new DateTime(fattura_cliente.data_creazione.Year, fattura_cliente.data_creazione.Month, fattura_cliente.data_creazione.Day),
                    number: null,
                    numeration: fattura_cliente.progr + "/" + fattura_cliente.sezionale.Trim(),
                    subject: "",
                    visibleSubject: "",
                    currency: new Currency(
                        id: "EUR"
                    ),
                    language: new Language(
                        code: "it",
                        name: "italiano"
                    ),
                    notes: note,
                    paymentMethod: new PaymentMethod(
                        id: int.Parse(id_metodo_pag)
                    ),
                    

                itemsList: new List<IssuedDocumentItemsListItem> {
                    new IssuedDocumentItemsListItem(
                        productId: null,
                        code: "",
                        name: fattura_cliente.descrizione,
                        grossPrice: Convert.ToDecimal(Math.Round(fattura_cliente.totale, 2, MidpointRounding.AwayFromZero)),
                        netPrice: Convert.ToDecimal(Math.Round(fattura_cliente.imponibile, 2, MidpointRounding.AwayFromZero) ),
                        category: "",
                        discount: 0,
                        qty: 1,
                        vat: new VatType(
                            id: fattura_cliente.cod_iva
                        )
                    )
                },

                paymentsList: new List<IssuedDocumentPaymentsListItem> {
                    new IssuedDocumentPaymentsListItem(
                        amount: 122,
                        dueDate: new DateTime(dueDate.Year, dueDate.Month, dueDate.Day),
                        paidDate: null,
                        status: null, // IssuedDocumentStatus.NotPaid,
                        paymentAccount: new PaymentAccount(
                            id: int.Parse(id_metodo_pag)
                        )
                    )
                }

            );



                string jsonInvoice = invoice.ToJson();

                // Here we put our invoice in the request object
                CreateIssuedDocumentRequest createIssuedDocumentRequest = new CreateIssuedDocumentRequest(
                    data: invoice
                );

                string jsonInvoiceData = createIssuedDocumentRequest.ToJson();

                // Now we are all set for the final call
                // Create the invoice: https://github.com/fattureincloud/fattureincloud-csharp-sdk/blob/master/docs/IssuedDocumentsApi.md#createissueddocument
                try
                {
                    CreateIssuedDocumentResponse result = apiInstance.CreateIssuedDocument(companyId, createIssuedDocumentRequest);
                    if (result!=null && result.Data.Id.HasValue)
                    {

                    }

                }
                catch (ApiException e)
                {
                    Console.WriteLine("Exception when calling IssuedDocumentsApi.CreateIssuedDocument: " + e.Message);
                    Console.WriteLine("Status Code: " + e.ErrorCode);
                    Console.WriteLine(e.StackTrace);
                }




            }
            catch (Exception ex)
            {

            }
```

**Reply by jehol75** [reply]
{
  "type": "invoice",
  "entity": {
    "name": "azienda di test",
    "vat_number": "0123456789",
    "address_street": "via delle vie 12",
    "address_postal_code": "12345",
    "address_city": "BS",
    "address_province": "BS",
    "country": "Italia"
  },
  "numeration": "10/999",
  "date": "2022-08-23",
  "currency": {
    "id": "EUR"
  },
  "language": {
    "code": "it",
    "name": "italiano"
  },
  "subject": "",
  "visible_subject": "",
  "notes": "Grazie per la preferenza accordataci.È un piacere lavorare con voi al vostro progetto.Cordiali saluti, SINFONIA cooperativa sociale",
  "payment_method": {
    "id": 1352151
  },
  "items_list": [
    {
      "code": "",
      "name": "test",
      "category": "",
      "qty": 1.0,
      "net_price": 123.0,
      "gross_price": 150.06,
      "vat": {
        "id": 0
      },
      "discount": 0.0
    }
  ],
  "payments_list": [
    {
      "due_date": "2022-09-07",
      "amount": 122.0,
      "payment_account": {
        "id": 1352151
      }
    }
  ]
}

**Reply by emajo** [reply]
Ho provato il suo codice sostituendo tutti i valori del suo oggetto `fattura_cliente` e la chiamata viene eseguita correttamente restituendo errore `422 il totale dei pagamenti non corrisponde al totale da pagare` (corretto dato che il totale da pagare corrisponde a 150.06 e il totale dei pagamenti è 122), la mia ipotesi è che nel suo caso l'errore venga dato proprio nel recuperare i dati dall'oggetto `fattura_cliente` o chiamando altri metodi e non le venga mostrato l'errore a causa del catch vuoto che si trova nell'ultima parte del suo codice.

**Reply by jehol75** [reply]
https://user-images.githubusercontent.com/2175833/186166162-58e76717-6c51-4b29-810f-1bf47ad75cee.mp4

Rifaccio un altro video dove puoi verificare un' altra volta che i dati sono recuperati (se no come vi mandavo il json), non ci sono altri metodi ne eccezioni e il vs metodo viene correttamente invocato e restituisce sempre null (incorretto in ogni caso)
Il catch che vi ho mandato e' vuoto perche' se vi mandavo la classe di logging poi ovviamente non vi andava, sorvolando su questo che e' irrilevante e comunque non ci sono exception

Avete dei blocchi su geolocalizzazione?

**Reply by emajo** [reply]
No, non abbiamo blocchi sulla localizzazione, a questo punto potrebbe essere un problema di dipendenze che vanno in conflitto, le chiedo se può provare a fare una chiamata utilizzando un client http per vedere se riesce a raggiungere i nostri servizi con successo, potrebbe anche provare a fare una veloce prova creando un nuovo progetto vuoto e facendo una semplice chiamata come la list issued documents utilizzando la nostra sdk, per capire se effettivamente sia un problema di dipendenze.

Inoltre le chiedo se sta utilizzando l'ultima versione dell'sdk o una versione precedente

**Reply by jehol75** [reply]
Conflitti sulle dipendenze la vedo improbabile dato che nella stessa classe faccio altre 2 chiamate ai vostri servizi e funzionano

![image](https://user-images.githubusercontent.com/2175833/186218537-d3e8284c-1763-439d-87f3-f3fa8199a01e.png)

per la chiamata al client stesso discorso di quanto sopra, le chiamate in immagine funzionano

la veloce prova del progetto nuovo non credo serva, sempre per quanto sopra, dato che le altre chiamate (payment.-methods) funzioano, comuqnue provero' a chiamare la lista degli issued documents

la versione SDK e' di 2/3 settimane fa come da immagine

![image](https://user-images.githubusercontent.com/2175833/186219085-694dec8a-e650-4f26-965c-6fcf6b3e5296.png)

**Reply by jehol75** [reply]
Get issued document, stessa classe, stesso token, funziona

![image](https://user-images.githubusercontent.com/2175833/186302864-3bfa6c96-8e49-47f6-8ea4-4df29f9b9a2a.png)

**Reply by emajo** [reply]
Vedo che sta usando la versione 2.0.10 dell'sdk, potrebbe provare ad aggiornarla alla 2.0.12? Potrebbe essere quello il problema.

**Reply by jehol75** [reply]
Aggiornato alla 2.0.12 compila tutto ma poi lanciadolo da il seguente errore
Probabilmente, dato che ho fatto copia/incolla dalla documentazione, tale documentazione risulta obsoleta

![image](https://user-images.githubusercontent.com/2175833/186425992-7c891d33-4492-4745-bb86-a7948230dac4.png)

**Reply by emajo** [reply]
Può mettere il codice del punto in cui le da errore?

**Reply by jehol75** [reply]
Va in errore nell'invocazione del metodo, cosa oltretutto strana... 

![image](https://user-images.githubusercontent.com/2175833/186680837-8c343f0b-1a60-4bf9-ac68-b0541744f24c.png)

Funziona anche chiamandola con client http mi dicevate giusto?

**Reply by emajo** [reply]
Difficile capire cosa c'è di sbagliato, bisognerebbe vedere dove tira effettivamente l'eccezione all'interno di InsertFattura, per quanto riguarda l'utilizzo di un client http è assolutamente possibile essendo la nostra una standard [api rest](https://developers.fattureincloud.it/docs/general-knowledge/rest-apis).

**Reply by jehol75** [reply]
Ripeto, senza caps lock : ) dentro l' insert fattura non ci entra.
Facendo anche F11 (step inside) non ci entra
Nel momento in cui viene invocata insert fattura lancia l' eccezione senza entraci
Il perche' non lo so
Ma dentro insert fattura - non ci entra - lancia l' eccezione riportata in immagine nel momento in cui viene invacata insert fattura
Mi son fatto capire? Serve un video?
Quindi bisogna gettare la spugna anche dopo questo aggiornamento?

**Reply by emajo** [reply]
Buongiorno @jehol75, se non entra nel suo metodo non credo sia un problema dovuto alla nostra sdk, e comunque non avendo altre informazioni risulta difficile aiutarla a risolvere anche questo problema.

**Reply by jehol75** [reply]
Allora, prima entrava in insert fattura, aggiornata come richiesto da voi alla nuova versione , senza cambiare una riga, da l' errore riportato sopra che dall' exception e' evidente che sia un vs errore.
Adesso, sta cosa sta evidentamente andando eccessivamente per le lunghe, ricordandovi che rappresento un cliente pagante e non e' un progetto free openSource ci si aspetterebbe un' assistenza adeguata e  un "non credo sia un problema dovuto alla nostra sdk," dato che il problema e' stato riportato adeguatamente
Il codice di sto SDK e' disponibile? Che nel caso vedo di risolvermela da solo

**Reply by emajo** [reply]
Come le ho già detto, la nostra SDK è utilizzata da molte altre integrazioni senza nessun problema, [qui](https://github.com/fattureincloud/fattureincloud-csharp-sdk) il codice della libreria.

**Reply by jehol75** [reply]
Va tutto senza problemi...alto valore aggiunto alla community

**Reply by valmoz** [reply]
Buongiorno @jehol75 
Sono il responsabile del team API di Fatture in Cloud. Mi spiace che lei abbia dovuto affrontare delle difficoltà nella fase di integrazione con le nostre API v2. 

Vorrei chiederle se è riuscito a procedere con l'integrazione o se si stanno verificando ulteriori problematiche per cui potrebbe essere necessario il nostro supporto.

Le auguro ancora una buona giornata.
Mauro

**Reply by jehol75** [reply]
Salve,
   mi sono scaricato il vostro codice per capire come mai la response e' null.

A seguito il motivo : nonostante ci sia un errorText l' exception non viene creata = response null

Constatato cio' sa darmi un motivo per cui ho l' errore : "Connessione sottostante chiusa: Errore imprevisto durante un'operazione di invio.."

![image](https://user-images.githubusercontent.com/2175833/190905417-388d2564-9a7f-48d1-becd-13738ebe9734.png)

![image](https://user-images.githubusercontent.com/2175833/190906862-5b877b35-c3c7-425c-94af-8dd0355db5eb.png)

**Reply by emajo** [reply]
Ciao @jehol75 sembrerebbe un errore legato a RestSharp che si presenta in alcune versioni .NET, dovrebbe poter risolvere aggiungendo la seguente linea di codice prima di eseguire la chiamata verso le nostre API 
`ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11;` [discussione stackoverflow](https://stackoverflow.com/questions/30409048/restsharp-tls-1-1).

Per migliorare la gestione degli errori della nostra sdk può aggiungere il seguente codice: 
```
            var apiInstance = new IssuedDocumentsApi(config);
            apiInstance.ExceptionFactory = (methodName, response) =>
            {
                var status = (int)response.StatusCode;
                if (status >= 400)
                {
                    return new ApiException(status,
                        string.Format("Error calling {0}: {1}", methodName, response.RawContent),
                        response.RawContent, response.Headers);
                }
                return new Exception(string.Format("Generic error: {0}", response.ErrorText));
            };
```
valuteremo se includerlo nel prossimo rilascio,

Emanuele

**Reply by jehol75** [reply]
Si avevo traovato il thread su restSharp, grazie

Un' altra cosa, ho messo i 2 flag per Einvoice 

![image](https://user-images.githubusercontent.com/2175833/191220246-36c46f9a-6505-4791-ad71-3d886e6a9b25.png)


Ma nelle lista e' segnata come "non elettrobnica" che manca?

![image](https://user-images.githubusercontent.com/2175833/191220784-2554deb5-37ee-4316-9df5-e5b2ed8a0290.png)

**Reply by valmoz** [reply]
Buongiorno,
ci faccia sapere se il thread su RestSharp le ha permesso di risolvere il problema.

Per quanto riguarda la e-invoice, lei ha settato il campo a true all'interno della _entity_, mentre dovrebbe essere nel corpo principale dell'_Issued Document_.

Può guardare il [punto tre della guida alla creazione delle fatture](https://developers.fattureincloud.it/docs/guides/invoice-creation#3%EF%B8%8F%E2%83%A3-step-three-e-invoice) per maggiori informazioni.

N.B.: si ricordi di impostare anche il campo _eiData_

Restiamo in attesa di ulteriori domande.

Grazie
Mauro

**Reply by jehol75** [reply]
Si il thread mi ha permesso di risolvere il problema

https://github.com/restsharp/RestSharp/issues/942

Messo a posto anche il flag e-invoice se anche faccio notare che la documentazione e' fuorviante

![image](https://user-images.githubusercontent.com/2175833/191244688-e0dec28e-31ec-4bbd-be5c-2b15fec25655.png)


Ultima domanda (spero)

Ho chiamato il sendEmail e l' email e' arrivata 

![image](https://user-images.githubusercontent.com/2175833/191244818-46c152c1-2220-4fa1-bfa7-8e92b505172d.png)


Ma sul portale la segna come "non inviata"

![image](https://user-images.githubusercontent.com/2175833/191244945-c4a736b4-d5ac-428b-903b-c09cb85dc306.png)

**Reply by emajo** [reply]
Ciao @jehol75 l'icona accanto a `non inviata` indica che la mail è stata mandata al cliente, mentre `non inviata` si riferisce allo stato della fattura elettronica, non inviata appunto all'SDI, per inviarla basterà utilizzare il metodo [send E-invoice](https://developers.fattureincloud.it/docs/guides/e-invoice-management/#-send-the-e-invoice).

**Reply by valmoz** [reply]
Buongiorno @jehol75 ha ancora delle domande relative all'utilizzo delle API?

Grazie
Mauro

**Comment by valmoz** [comment, marked as answer]
Riassumo per comodità.

L'errore null sulla C# SDK sembrerebbe essere legato ad un bug sulla libreria RestSharp, che si presenta in alcune versioni .NET.

Per risolvere è possibile aggiungere la seguente linea di codice prima di eseguire la chiamata verso le nostre API:
```cs
ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12 | SecurityProtocolType.Tls11; 
```

[discussione stackoverflow](https://stackoverflow.com/questions/30409048/restsharp-tls-1-1).

Per migliorare la gestione degli errori della nostra sdk è possibile aggiungere il seguente codice:
```cs
var apiInstance = new IssuedDocumentsApi(config);
apiInstance.ExceptionFactory = (methodName, response) =>
{
    var status = (int)response.StatusCode;
    if (status >= 400)
    {
        return new ApiException(status,
            string.Format("Error calling {0}: {1}", methodName, response.RawContent),
            response.RawContent, response.Headers);
    }
    if (status == 0)
    {
        return new ApiException(status,
            string.Format("Error calling {0}: {1}", methodName, response.ErrorText), response.ErrorText);
    }            
    return null;
};
```
valuteremo se modificare la exception factory nei prossimi rilasci.

Grazie
Mauro

**Reply by francescoffz** [reply]
Quel codice solleva sempre eccezione, anche in caso di richieste andate a buon fine, con status code 200.

Così invece funziona:

apiInstance.ExceptionFactory = (methodName, response) =>
{
    var status = (int)response.StatusCode;
    if (status >= 400)
    {
        return new ApiException(status,
            string.Format("Error calling {0}: {1}", methodName, response.RawContent),
            response.RawContent, response.Headers);
    }
    return null;
};

**Reply by emajo** [reply]
Grazie @francescoffz effettivamente è errato, procediamo a modificare l'esempio.
